src/system.cpp

```diff
@@ -760,16 +760,31 @@ int System::CastOffPages(FunctorParams *functorParams)
     CastOffPagesParams *params = dynamic_cast<CastOffPagesParams *>(functorParams);
     assert(params);
 
-    int currentShift = params->m_shift;
+    int headerAndFooterHeight = 0;
     // We use params->m_pageHeadHeight to check if we have passed the first page already
     if (params->m_pgHeadHeight != VRV_UNSET) {
-        currentShift += params->m_pgHeadHeight + params->m_pgFootHeight;
+        headerAndFooterHeight = params->m_pgHeadHeight + params->m_pgFootHeight;
     }
     else {
-        currentShift += params->m_pgHead2Height + params->m_pgFoot2Height;
+        headerAndFooterHeight = params->m_pgHead2Height + params->m_pgFoot2Height;
     }
 
-    if ((params->m_currentPage->GetChildCount() > 0) && (this->m_drawingYRel - this->GetHeight() - currentShift < 0)) {
+    const auto systemMaxPerPage = params->m_doc->GetOptions()->m_systemMaxPerPage.GetValue();
+    bool placeOnNewPage = systemMaxPerPage && systemMaxPerPage == params->m_currentPage->GetChildCount();
+    if (!placeOnNewPage && systemMaxPerPage) {
+        int systemsTotalHeight = 0;
+        for (const auto child : *params->m_currentPage->GetChildren()) {
+            const auto childSystem = dynamic_cast<System *>(child);
+            assert(childSystem);
+
+            systemsTotalHeight += childSystem->GetHeight();
+        }
+
+        placeOnNewPage = systemsTotalHeight + this->GetHeight() > params->m_pageHeight - headerAndFooterHeight;
+    }
+
+    const auto currentShift = params->m_shift + headerAndFooterHeight;
+    if (placeOnNewPage || (this->m_drawingYRel - this->GetHeight() - currentShift < 0)) {
```

Comment:
Again, I don't think you need to change that much. If we need to fix footer and header spacing, please fix is separately.

Reply:
That is not about footer and header. Since we need to place not more systems than systemMaxPerPage I have to calculate how match space left on the page to decide place it on new page or not.

Reply:
This is what I do not understand. The standard algorithm already does this. You only need to add one test in the condition, namely to check if the option is set and if the maximum number of system on the page is already reached.

Reply:
Ok. I see the issue. Looks like castOfDoc first calls the LayOutVertically for the page which contains all systems of the document what is break logic of margin calculation and I have incorrect drawingYRel offset, as result systems overlapping and I had to calculate system height in place. 

It looks like I need somehow to distinguish that case and return 0 as a system margin in all cases when the maxPerPage is set and we do not do CastOffPages... In that case I will not need any additional calculation in that method.

Is there any simple way to distinguish that case? Could you please advice?

Reply:
Here: https://github.com/rism-ch/verovio/blob/df5c905d131b887730af9e7ec1afb1bf2543b8ec/src/system.cpp#L784
one additional test to check also if you have reached the maximum number of system on the current page. That should be it

Reply:
See this https://github.com/rism-ch/verovio/commit/b99f57b3adfdd413b3ed799535ebba5fb0f24b47 
With a value of 2 (as coded) here it produces this - but renders normally with 0 or with a value higher than what fits on the page (e.g. 10)
Please make your PR with the code for adding the option itself to this branch (develop-max-system)

![image](https://user-images.githubusercontent.com/689412/85104478-fc43b680-b208-11ea-9927-1a464eab4282.png)

Reply:
Hi @lpugin , it looks ok when all systems can fit the page, but when it's not... the systems overlapping each other, since initial margin calculated incorrectly and drawingYRel are shifted. 

I see the following picture:
![image](https://user-images.githubusercontent.com/11978860/85117929-e5f52500-b21f-11ea-8cea-e7f6a94b540a.png)

So, my idea is to set initial system margins to 0 during first LayOutVertically for the one page document, and them it should make proper calculation when layout on smaller pages. The problem for me now - is to distinguish this cases.

Reply:
Perhaps, you do not really understand what I'm trying to do here... but the tricky part is margins, since first system should be alligned to the top of page, and the last one to the bottom. Margins should be equal in the middle, here is the sample (except last page):
![image](https://user-images.githubusercontent.com/11978860/85118778-15f0f800-b221-11ea-987f-b01aa75bfd68.png)

Reply:
How id you produce this example?? It this the develop-max-system branch? Could you post the file and the parameter you used? 
Sorry, I do not understand what you are trying to achieve. If the content does not fit on a page with --spacing-staff and --spacing-system set to 0, then I do not understand how you are going to make it fit on the page. This is why the example implementation I gave will not put the maximum number of system on a page if they do not fit. I don't understand how it yielded the overlap you a showing here.
Why can't you simply have --spacing-staff and --spacing-system to 0 and then justify the content of the page vertically with --justify-vertically (with our without --justify-systems-only)?

Reply:
The mei file is this one: https://github.com/music-encoding/sample-encodings/blob/master/MEI_4.0/Music/Complete_examples/Grieg_op.43_butterfly.mei 
with the following parameters:
verovio -a --system-max-per-page 10 -r /work/verovio/data /work/Samples/Grieg_op.43_butterfly.mei

I'm using forked version as base, so I'll try cherry pick my commit on the top of rism develop and double check, I'll check your suggestion as well.